bpf/bpf_host.c

```diff
@@ -652,21 +730,63 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 }
 
 static __always_inline int
-tail_handle_ipv4(struct __ctx_buff *ctx, __u32 ipcache_srcid, const bool from_host)
+tail_handle_ipv4_cont(struct __ctx_buff *ctx, bool from_host)
 {
 	__u32 proxy_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
 	int ret;
 	__s8 ext_err = 0;
 
 	ctx_store_meta(ctx, CB_SRC_LABEL, 0);
 
-	ret = handle_ipv4(ctx, proxy_identity, ipcache_srcid, from_host, &ext_err);
+	ret = handle_ipv4_cont(ctx, proxy_identity, from_host, &ext_err);
 	if (IS_ERR(ret))
 		return send_drop_notify_error_ext(ctx, proxy_identity, ret, ext_err,
 						  CTX_ACT_DROP, METRIC_INGRESS);
 	return ret;
 }
 
+declare_tailcall_if(is_defined(ENABLE_HOST_FIREWALL), CILIUM_CALL_IPV4_CONT_FROM_HOST)
+int tail_handle_ipv4_cont_from_host(struct __ctx_buff *ctx)
+{
+	return tail_handle_ipv4_cont(ctx, true);
+}
+
+declare_tailcall_if(is_defined(ENABLE_HOST_FIREWALL), CILIUM_CALL_IPV4_CONT_FROM_NETDEV)
+int tail_handle_ipv4_cont_from_netdev(struct __ctx_buff *ctx)
+{
+	return tail_handle_ipv4_cont(ctx, false);
+}
+
```

Comment:
These are potentially executed as tail-calls, so we need to check for errors and call `send_drop_notify_error()`.

Reply:
Missed that, thanks!

Note that the situation has slightly changed vs current code. In current code `tail_handle_ipv4_from_{host,netdev}()` are unconditional tail-calls, so we know that they *have* to handle drop/trace notifications.

With `declare_tailcall_if()` we also need to consider that the function gets inlined, and the surrounding code then expects to handle drop/trace notifications. Not a problem for drop notifications (as they are currently implemented through a tail-call, so control will be gone after the first drop notify). But trace notifications would get duplicated.

Reply:
> I'd keep it as is then, at least for now. We also have a similar pattern in bpf_lxc.

oh yeah absolutely. I've been wanting to clean/fix some drop notification things anyway, maybe I'll tackle it along then at that point.